writing_state.go

```diff
@@ -0,0 +1,128 @@
+package dastard
+
+import (
+	"bufio"
+	"fmt"
+	"os"
+	"sync"
+	"time"
+)
+
+// WritingState monitors the state of file writing.
+type WritingState struct {
+	Active                            bool
+	Paused                            bool
+	BasePath                          string
+	FilenamePattern                   string
+	experimentStateFile               *os.File
+	ExperimentStateFilename           string
+	ExperimentStateLabel              string
+	ExperimentStateLabelUnixNano      int64
+	ExternalTriggerFilename           string
+	externalTriggerNumberObserved     int
+	externalTriggerFileBufferedWriter *bufio.Writer
+	externalTriggerTicker             *time.Ticker
+	externalTriggerFile               *os.File
+	sync.Mutex
+}
+
+// IsActive will return ws.Active, with propert locking
+func (ws *WritingState) IsActive() bool {
+	ws.Lock()
+	defer ws.Unlock()
+	return ws.Active
+}
+
+// ComputeState will return a property-by-property copy of the WritingState.
+// It will not copy the "active" features like open files, tickers, etc.
+func (ws *WritingState) ComputeState() WritingState {
+	ws.Lock()
+	defer ws.Unlock()
+	var copyState WritingState
+	copyState.Active = ws.Active
+	copyState.Paused = ws.Paused
+	copyState.BasePath = ws.BasePath
+	copyState.FilenamePattern = ws.FilenamePattern
+	copyState.ExperimentStateFilename = ws.ExperimentStateFilename
+	copyState.ExperimentStateLabel = ws.ExperimentStateLabel
+	copyState.ExperimentStateLabelUnixNano = ws.ExperimentStateLabelUnixNano
+	copyState.ExternalTriggerFilename = ws.ExternalTriggerFilename
+	copyState.externalTriggerNumberObserved = ws.externalTriggerNumberObserved
+	return copyState
+}
+
+// Start will set the WritingState to begin writing
+func (ws *WritingState) Start(filenamePattern, path string) error {
+	ws.Lock()
+	defer ws.Unlock()
+	ws.Active = true
+	ws.Paused = false
+	ws.BasePath = path
+	ws.FilenamePattern = filenamePattern
+	ws.ExperimentStateFilename = fmt.Sprintf(filenamePattern, "experiment_state", "txt")
+	ws.ExternalTriggerFilename = fmt.Sprintf(filenamePattern, "external_trigger", "bin")
+	return ws.setExperimentStateLabel(time.Now(), "START")
+}
+
+// Stop will set the WritingState to be completely stopped
+func (ws *WritingState) Stop() error {
+	ws.Lock()
+	defer ws.Unlock()
+	ws.Active = false
+	ws.Paused = false
+	ws.FilenamePattern = ""
+	ws.setExperimentStateLabel(time.Now(), "STOP")
```

Comment:
check errors

Reply:
Good catch. Actually, tests fail if you check for and return any non-nil error returned by ws.setExperimentStateLabel in line 74. I did not figure out whether this was a mistake in the test, because it was easier to ignore the return value. But you're right, that's not the right approach. Let me see if the test is in error....